src/System.Security.Cryptography.Pkcs/tests/EnvelopedCms/KeyAgreeRecipientInfoTests.cs

```diff
@@ -0,0 +1,354 @@
+
+            KeyAgreeRecipientInfo recipient = FixedValueKeyAgree1();
+            byte[] encryptedKey = recipient.EncryptedKey;
+            Assert.Equal<byte>(expectedEncryptedKey, encryptedKey);
+        }
+
+        [Fact]
+        public static void TestKeyAgreeOriginatorIdentifierOrKey_RoundTrip()
+        {
+            KeyAgreeRecipientInfo recipient = EncodeKeyAgreel();
+            SubjectIdentifierOrKey originator = recipient.OriginatorIdentifierOrKey;
+            Assert.Equal(SubjectIdentifierOrKeyType.PublicKeyInfo, originator.Type);
+            object value = originator.Value;
+            Assert.True(value is PublicKeyInfo);
+            PublicKeyInfo pki = (PublicKeyInfo)value;
+            AlgorithmIdentifier a = pki.Algorithm;
+            Assert.Equal(Oids.Dh, a.Oid.Value);
+            byte[] key = pki.KeyValue;
+            Assert.NotNull(key);  // Key is randomly generated (and encoding makes the length unpredictable) so not much we can do here.
+        }
+
+        [Fact]
+        public static void TestKeyAgreeOriginatorIdentifierOrKey_FixedValue()
+        {
+            KeyAgreeRecipientInfo recipient = FixedValueKeyAgree1();
+            SubjectIdentifierOrKey originator = recipient.OriginatorIdentifierOrKey;
+            Assert.Equal(SubjectIdentifierOrKeyType.PublicKeyInfo, originator.Type);
+            object value = originator.Value;
+            Assert.True(value is PublicKeyInfo);
+            PublicKeyInfo pki = (PublicKeyInfo)value;
+            AlgorithmIdentifier a = pki.Algorithm;
+            Assert.Equal(Oids.Dh, a.Oid.Value);
+            byte[] key = pki.KeyValue;
+            byte[] expectedKey =
+                ("0281806F96EF8C53A6919CC976E88B8F426696E7B7970ABC6BD4ABBDCF4CF34F89CEB6E8EF675000FAD2ECA3CAF9D0E51B00"
+                + "4FD19A943F1779748F343FE2059E6E8208D64CB2A5BF33B2C41C20F4AE950D8F8BD720F5747D7930AF86C612088747B5315A"
+                + "E68159A5AE8A80E928AA71F4E889CB2D581845EDC8F79DA5894CB7A40F9FBE").HexToByteArray();
+
+            Assert.Equal(expectedKey, key);
+        }
+
+        [Fact]
+        public static void TestKeyAgreeDate_RoundTrip()
+        {
+            KeyAgreeRecipientInfo recipient = EncodeKeyAgreel();
+            DateTime ignore;
+            Assert.Throws<InvalidOperationException>(() => ignore = recipient.Date);
+        }
+
+        [Fact]
+        public static void TestKeyAgreeDate_FixedValue()
+        {
+            KeyAgreeRecipientInfo recipient = FixedValueKeyAgree1();
+            DateTime ignore;
+            Assert.Throws<InvalidOperationException>(() => ignore = recipient.Date);
+        }
+
+        [Fact]
+        public static void TestKeyAgreeDate_RoundTrip_Ski()
+        {
+            KeyAgreeRecipientInfo recipient = EncodeKeyAgreel(SubjectIdentifierType.SubjectKeyIdentifier);
+            DateTime date = recipient.Date;
```

Comment:
From MSDN:
Starting with the .NET Framework version 2.0, the return value [of FromFileTimeUtc()] is a DateTime<https://msdn.microsoft.com/en-us/library/system.datetime(v=vs.110).aspx> whose Kind<https://msdn.microsoft.com/en-us/library/system.datetime.kind(v=vs.110).aspx> property is Utc<https://msdn.microsoft.com/en-us/library/shx7s921(v=vs.110).aspx>.

So unless someone managed to dig up a copy of 1.0 and somehow get this stuff to run on that, this implementation will always return a UTC DateTime.

Having said that, the docs for KeyAgreeRecipientInfo.Date don’t actually guarantee the DateTimeKind.